### PR TITLE
 Add task relocation Postgresql persistence store implementation

### PR DESCRIPTION
**Pull request by tbak**

### CODE REVIEW

File: titus-common/src/main/java/com/netflix/titus/common/util/DateTimeExt.java [hunk 1]
```diff
@@ -93,4 +114,24 @@ public static String toTimeUnitString(long timeMs) {
         }
         return sb.toString();
     }
+
+    /**
+     * Given an interval, and a number of items generated per interval, create string representation of the
+     * corresponding rate. For example, interval=10sec, itemsPerInterval=5 gives rate="0.5 ???/sec".
```
Comment (corindwyer) [inline titus-common/src/main/java/com/netflix/titus/common/util/DateTimeExt.java]: ??? -> items

File: titus-server-master/src/main/java/com/netflix/titus/master/eviction/service/quota/job/TaskRelocationLimitController.java [hunk 2]
```diff
@@ -79,7 +94,31 @@ public long getQuota() {
             }
         }
 
-        return quota;
+        return quota > 0
+                ? quotaBuilder.withQuota(quota).withMessage("Per task limit is %s", perTaskLimit).build()
+                : quotaBuilder.withQuota(0).withMessage("All tasks of the job reached its maximum eviction limit %s", perTaskLimit).build();
```
Comment (corindwyer) [inline titus-server-master/src/main/java/com/netflix/titus/master/eviction/service/quota/job/TaskRelocationLimitController.java]: I think this message needs to be rephrased.

File: titus-server-master/src/main/java/com/netflix/titus/master/eviction/service/quota/system/SystemQuotaController.java [hunk 3]
```diff
@@ -94,6 +97,10 @@ private void resetDisruptionBudget(SystemDisruptionBudget newDisruptionBudget) {
         this.disruptionBudget = newDisruptionBudget;
         this.systemTokenBucket = newTokenBucket(newDisruptionBudget);
         this.inTimeWindowPredicate = TimeWindowFunctions.isInTimeWindowPredicate(titusRuntime, disruptionBudget.getTimeWindows());
+        this.quotaMessage = String.format("System quota token bucket: capacity=%s, refillRate=%s",
```
Comment (corindwyer) [inline titus-server-master/src/main/java/com/netflix/titus/master/eviction/service/quota/system/SystemQuotaController.java]: Is refillRate an actual rate or the name of the strategy?

File: titus-supplementary-component/task-relocation/src/main/java/com/netflix/titus/supplementary/relocation/workflow/DefaultRelocationWorkflowExecutor.java [hunk 4]
```diff
@@ -183,7 +183,7 @@ public void shutdown() {
 
     private void nextRelocationStep(ExecutionContext executionContext) {
         long count = executionContext.getExecutionId().getTotal();
-        boolean descheduling = lastDeschedulingTimestamp + configuration.getDeschedulingIntervalMs() > titusRuntime.getClock().wallTime();
+        boolean descheduling = lastDeschedulingTimestamp + configuration.getDeschedulingIntervalMs() <= titusRuntime.getClock().wallTime();
```
Comment (corindwyer) [inline titus-supplementary-component/task-relocation/src/main/java/com/netflix/titus/supplementary/relocation/workflow/DefaultRelocationWorkflowExecutor.java]: make a utility function? I have some private ones in some places but maybe a general purpose one would reduce this type of mistake.

File: titus-ext/jooq/src/main/resources/relocation-schema.xml [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        </table>
+    </tables>
+
+    <columns>
+        <!-- relocation_plan -->
+        <column>
+            <table_schema>relocation</table_schema>
+            <table_name>relocation_plan</table_name>
+            <column_name>task_id</column_name>
+            <data_type>VARCHAR</data_type>
```
[the remark this answers is not in the document]
Comment (tbak) [reply titus-ext/jooq/src/main/resources/relocation-schema.xml]: As discussed offline, lets not use database specific extensions yet.

File: titus-ext/jooq/src/main/resources/relocation-schema.xml [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+            <data_type>BIGINT</data_type>
+            <is_nullable>false</is_nullable>
+        </column>
+
+        <!-- relocation_status -->
+        <column>
+            <table_schema>relocation</table_schema>
+            <table_name>relocation_status</table_name>
+            <column_name>task_id</column_name>
+            <data_type>VARCHAR</data_type>
```
[the remark this answers is not in the document]
Comment (tbak) [reply titus-ext/jooq/src/main/resources/relocation-schema.xml]: As discussed offline, lets not use database specific extensions yet.

File: titus-supplementary-component/task-relocation/src/main/java/com/netflix/titus/supplementary/relocation/workflow/DefaultRelocationWorkflowExecutor.java [hunk 7]
```diff
@@ -63,6 +64,8 @@
 
     private static final long STALENESS_THRESHOLD_MS = 30_000;
 
+    private static final Map<String, TaskRelocationPlan> PLANS_NOT_READY = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (tbak) [reply titus-supplementary-component/task-relocation/src/main/java/com/netflix/titus/supplementary/relocation/workflow/DefaultRelocationWorkflowExecutor.java]: No, this object must have a unique value. `EmptyMap` may appear later and be a valid subsequent state. I will add a comment.

File: titus-ext/jooq/src/main/resources/relocation-schema.xml [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+            <table_name>relocation_plan</table_name>
+            <column_name>task_id</column_name>
+            <data_type>VARCHAR</data_type>
+            <character_maximum_length>64</character_maximum_length>
+            <is_nullable>false</is_nullable>
+        </column>
+        <column>
+            <table_schema>relocation</table_schema>
+            <table_name>relocation_plan</table_name>
+            <column_name>reason</column_name>
```
[the remark this answers is not in the document]
Comment (tbak) [reply titus-ext/jooq/src/main/resources/relocation-schema.xml]: +1

File: titus-ext/jooq/src/main/resources/relocation-schema.xml [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+            <data_type>VARCHAR</data_type>
+            <character_maximum_length>64</character_maximum_length>
+            <is_nullable>false</is_nullable>
+        </column>
+        <column>
+            <table_schema>relocation</table_schema>
+            <table_name>relocation_plan</table_name>
+            <column_name>reason_message</column_name>
+            <data_type>VARCHAR</data_type>
```
[the remark this answers is not in the document]
Comment (tbak) [reply titus-ext/jooq/src/main/resources/relocation-schema.xml]: +1

File: titus-ext/jooq/src/main/resources/relocation-schema.xml [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+            <character_maximum_length>1024</character_maximum_length>
+            <is_nullable>false</is_nullable>
+        </column>
+        <column>
+            <table_schema>relocation</table_schema>
+            <table_name>relocation_plan</table_name>
+            <column_name>relocation_time</column_name>
+            <data_type>BIGINT</data_type>
```
[the remark this answers is not in the document]
Comment (tbak) [reply titus-ext/jooq/src/main/resources/relocation-schema.xml]: +1

File: titus-ext/jooq/src/main/resources/relocation-schema.xml [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+            <data_type>VARCHAR</data_type>
+            <character_maximum_length>64</character_maximum_length>
+            <is_nullable>false</is_nullable>
+        </column>
+        <column>
+            <table_schema>relocation</table_schema>
+            <table_name>relocation_status</table_name>
+            <column_name>relocation_state</column_name>
+            <data_type>VARCHAR</data_type>
+            <character_maximum_length>16</character_maximum_length>
```
[the remark this answers is not in the document]
Comment (tbak) [reply titus-ext/jooq/src/main/resources/relocation-schema.xml]: This is enum name, so it should not be too long, but I agree that 64 is safer.

File: titus-ext/jooq/src/main/resources/relocation-schema.xml [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+            <column_name>status_code</column_name>
+            <data_type>VARCHAR</data_type>
+            <character_maximum_length>64</character_maximum_length>
+            <is_nullable>false</is_nullable>
+        </column>
+        <column>
+            <table_schema>relocation</table_schema>
+            <table_name>relocation_status</table_name>
+            <column_name>status_message</column_name>
+            <data_type>VARCHAR</data_type>
```
[the remark this answers is not in the document]
Comment (tbak) [reply titus-ext/jooq/src/main/resources/relocation-schema.xml]: +1

File: titus-ext/jooq/src/main/resources/relocation-schema.xml [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+            <column_name>reason_code</column_name>
+            <data_type>VARCHAR</data_type>
+            <character_maximum_length>64</character_maximum_length>
+            <is_nullable>false</is_nullable>
+        </column>
+        <column>
+            <table_schema>relocation</table_schema>
+            <table_name>relocation_status</table_name>
+            <column_name>reason_message</column_name>
+            <data_type>VARCHAR</data_type>
```
[the remark this answers is not in the document]
Comment (tbak) [reply titus-ext/jooq/src/main/resources/relocation-schema.xml]: +1

File: titus-ext/jooq/src/main/resources/relocation-schema.xml [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+            <character_maximum_length>1024</character_maximum_length>
+            <is_nullable>false</is_nullable>
+        </column>
+        <column>
+            <table_schema>relocation</table_schema>
+            <table_name>relocation_status</table_name>
+            <column_name>relocation_time</column_name>
+            <data_type>BIGINT</data_type>
```
[the remark this answers is not in the document]
Comment (tbak) [reply titus-ext/jooq/src/main/resources/relocation-schema.xml]: +1

File: titus-common/src/main/java/com/netflix/titus/common/util/DateTimeExt.java (hunk 1 above)
Comment (tbak) [reply titus-common/src/main/java/com/netflix/titus/common/util/DateTimeExt.java]: +1

File: titus-ext/jooq/src/main/java/com/netflix/titus/ext/jooq/relocation/JooqTaskRelocationResultStore.java [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+            CompletionStage<Void> asyncAction = DSL.using(dslContext.configuration())
+                    .transactionAsync(configuration -> {
+                                loadToCache(findNotCached(taskRelocationStatuses), configuration);
+
+                                List<StoreQuery<JRelocationStatusRecord>> queries = taskRelocationStatuses.stream()
+                                        .map(this::newCreateOrUpdateQuery)
+                                        .collect(Collectors.toList());
+
+                                configuration.dsl()
+                                        .batch(queries)
```
[the remark this answers is not in the document]
Comment (tbak) [reply titus-ext/jooq/src/main/java/com/netflix/titus/ext/jooq/relocation/JooqTaskRelocationResultStore.java]: We could improve this in the future, but this should not be a problem today.

[the remark this answers is not in the document]
Comment (tbak) [reply titus-ext/jooq/src/main/java/com/netflix/titus/ext/jooq/relocation/JooqTaskRelocationResultStore.java]: I have added unit test to persist 10K values in a batch. It works with the embedded `hsqldb`. I will have to test this manually with postgres.

File: titus-server-master/src/main/java/com/netflix/titus/master/eviction/service/quota/system/SystemQuotaController.java (hunk 3 above)
Comment (tbak) [reply titus-server-master/src/main/java/com/netflix/titus/master/eviction/service/quota/system/SystemQuotaController.java]: Renamed to `refillStrategy`

File: titus-supplementary-component/task-relocation/src/main/java/com/netflix/titus/supplementary/relocation/endpoint/TaskRelocationPlanPredicate.java [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+        }
+    }
+
+    @Override
+    public boolean test(TaskRelocationPlan relocationPlan) {
+        return predicate.test(relocationPlan);
+    }
+
+    private Optional<Predicate<TaskRelocationPlan>> newJobIdsPredicate(String jobIds) {
+        List<String> idList = StringExt.splitByComma(jobIds);
```
[the remark this answers is not in the document]
Comment (tbak) [reply titus-supplementary-component/task-relocation/src/main/java/com/netflix/titus/supplementary/relocation/endpoint/TaskRelocationPlanPredicate.java]: The `splitByComma` method returns list, as the order may be important in some cases (not here). The performance overhead of doing list to set conversion is negligible here.

File: titus-supplementary-component/task-relocation/src/main/java/com/netflix/titus/supplementary/relocation/endpoint/TaskRelocationPlanPredicate.java [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+        if (idList.isEmpty()) {
+            return Optional.empty();
+        }
+        Set<String> idSet = new HashSet<>(idList);
+
+        return Optional.of(newJobPredicate(job -> idSet.contains(job.getId())));
+    }
+
+    private Optional<Predicate<TaskRelocationPlan>> newTaskIdsPredicate(String taskIds) {
+        List<String> idList = StringExt.splitByComma(taskIds);
```
[the remark this answers is not in the document]
Comment (tbak) [reply titus-supplementary-component/task-relocation/src/main/java/com/netflix/titus/supplementary/relocation/endpoint/TaskRelocationPlanPredicate.java]: See the comment above.

File: titus-supplementary-component/task-relocation/src/main/java/com/netflix/titus/supplementary/relocation/workflow/DefaultRelocationWorkflowExecutor.java (hunk 4 above)
Comment (tbak) [reply titus-supplementary-component/task-relocation/src/main/java/com/netflix/titus/supplementary/relocation/workflow/DefaultRelocationWorkflowExecutor.java]: Good idea. I have added a `isPast` method to the `Clock` interface.

File: titus-server-master/src/main/java/com/netflix/titus/master/eviction/service/quota/job/TaskRelocationLimitController.java (hunk 2 above)
Comment (tbak) [reply titus-server-master/src/main/java/com/netflix/titus/master/eviction/service/quota/job/TaskRelocationLimitController.java]: Will try

**Pull request merged**